### PR TITLE
test(generators): add semantic assertions to generator contract tests

### PR DESCRIPTION
**Pull request by OliverWolffGIP**
## Summary

Replace the vacuous `assertNotNull` / `isEmpty` / `length > N` bodies in the 17 weak generator
contract tests with per-value defining-property assertions plus per-class aggregate-coverage
assertions, modelled on `SupportedValidationTypeGeneratorContractTest`. A shared
`GeneratorContractAssertions` helper holds the marker vocabularies (traversal, null-byte, CRLF,
control-char) so the pattern applies uniformly across url/, header/, encoding/, injection/, and
cookie/ generators instead of being copy-pasted per test.

## Changes

- `GeneratorContractAssertions.java` (new) — shared defining-property assertion helper: marker
  vocabularies plus pipeline round-trip helpers, consumed by every contract test below.
- 17 generator contract tests (url/, header/, encoding/, injection/, cookie/) gain per-value
  defining-property assertions and per-class aggregate-coverage assertions in place of the
  previous vacuous checks.
- `AllGeneratorsIntegrationTest.shouldHaveAllGeneratorsImplemented` now asserts real
  generator-output properties instead of `assertNotNull` on already-initialized fields.
- `AttackCookieGeneratorTest` / `ValidCookieGeneratorTest` — replaced ad-hoc field checks with
  record-equality assertions.
- `EncodingCombinationGenerator.java` (production fix) — corrects the single-level-encoding
  generation so the class actually emits the attack pattern its contract test now asserts against.
- `URLLengthLimitAttackGenerator.java` (production fix) — corrects the `Generators.strings(token, n, m)`
  usage so generated strings reliably exceed the configured length limit.
- `URLLengthLimitAttackTest.java` — activates the previously-silent `isPathBasedLengthAttack` guard
  path with an admitted-sample-count assertion, so the guard can no longer no-op invisibly.

## Absorbed from PLAN-12

Per operator decision (phase-2-refine Q1, phase-3-outline Q2), this plan absorbs two of PLAN-12's
planned generator production fixes as coordinated production+test pairs, landing each fix together
with its new semantic assertion in the same change set so no assertion is ever merged in a
known-failing state:

- `EncodingCombinationGenerator` — production fix + defining-property assertions (findings TQ-15, TQ-16)
- `URLLengthLimitAttackGenerator` — production fix + defining-property assertions (finding TQ-17)

This forfeits the original PLAN-11-must-land-before-PLAN-12 sequencing for these two generators
only; all other PLAN-12 generator fixes remain out of scope for this plan.

## Test Plan

- [ ] Verification command passed (`python3 .plan/execute-script.py plan-marshall:build-maven:maven run --command-args "verify -Ppre-commit"`)
- [ ] Manual testing completed (if applicable)

## Related Issues

---
Generated by plan-finalize skill

## Intent

**Problem**: 17 of 20 generator contract tests asserted nothing meaningful about the values they
generated (vacuous `assertNotNull` / `isEmpty` / `length > N` checks), and
`AllGeneratorsIntegrationTest.shouldHaveAllGeneratorsImplemented` could never fail. As a result, two
generators that do not actually emit their advertised attack pattern (`EncodingCombinationGenerator`,
`URLLengthLimitAttackGenerator`) went undetected by their own contract tests.

**Approach**: Introduce one shared defining-property assertion helper
(`GeneratorContractAssertions`) and apply it uniformly across every generator contract test —
per-value marker assertions plus per-class aggregate-coverage assertions, following the existing
`SupportedValidationTypeGeneratorContractTest` pattern. Where a new assertion exposed a real
generator defect, the production fix lands in the SAME change set as its assertion, never in a
follow-up, so no assertion is ever merged in a known-failing state. This absorbs
`EncodingCombinationGenerator` and `URLLengthLimitAttackGenerator` from PLAN-12 as one coordinated
pair, by explicit operator decision.

**Non-goals**: This plan does not fix any other PLAN-12-owned generator — that scope stays with
PLAN-12. It does not add a blanket pipeline round-trip assertion to every generator; that is applied
only where the contract is unambiguous, with `InvalidURLGenerator`,

_[Intent truncated — 1385 of 1608 characters shown; full outline in the plan workspace]_

## Summary by CodeRabbit

- **Bug Fixes**
  - Corrected URL encoding to prevent incorrectly double-encoded values.
  - Ensured generated URL length-limit attack cases consistently exceed strict limits.
- **Security Validation**
  - Strengthened validation coverage across paths, parameters, headers, cookies, URLs, encoding, Unicode, and request-smuggling scenarios.
  - Improved verification that malicious inputs are rejected and legitimate inputs are accepted.
- **Tests**
  - Added comprehensive coverage for attack variants, edge cases, output types, and generated-value diversity.
  - Added safeguards to detect ineffective or incomplete test runs.

### CODE REVIEW

File: cui-http-core/src/test/java/de/cuioss/http/security/generators/cookie/AttackCookieGeneratorTest.java [hunk 1]
```diff
@@ -19,37 +19,218 @@
...
+            "\t injected", "\r injected", "\n injected");
+
+    private static final List<String> HOSTILE_DOMAINS = List.of(
+            ".evil.com", ".attacker.net", ".malicious.org");
+    private static final String TRAVERSAL_PATH_ATTRIBUTE = "Path=../../../";
+    private static final String NEGATIVE_MAX_AGE_ATTRIBUTE = "Max-Age=-1";
+
+    /** The malformed-attribute branch emits exactly these four forms. */
+    private static final Set<String> MALFORMED_ATTRIBUTES = Set.of(
+            "Domain=; Path=", "Invalid=Attribute; Bad=Value", "Domain=", "=; Path=/");
```
[the remark this answers is not in the document]
Comment (OliverWolffGIP) [reply cui-http-core/src/test/java/de/cuioss/http/security/generators/cookie/AttackCookieGeneratorTest.java]: Accepted, out of scope for this plan. Sharing/deriving generator-family catalogs at test runtime across the cookie, boundary-fuzzing, path-traversal, header-injection, null-byte and valid-URL contract tests is a structural refactor spanning most of the generator suite (correctly labeled a heavy lift), not a narrow duplication fix. This plan's scope and lean simplicity posture cover per-generator defining-property assertions, not a shared-catalog architecture. The already-shared cross-cutting vocabularies (traversal/null-byte/CRLF/shell markers) live in GeneratorContractAssertions; the remaining literals are generator-specific payload text, not duplicated marker sets. Worth raising as a dedicated follow-up plan if the duplication proves costly in practice.

File: cui-http-core/src/test/java/de/cuioss/http/security/generators/cookie/ValidCookieGeneratorTest.java [hunk 2]
```diff
@@ -19,36 +19,143 @@
...
 
-        // Test equals and hashCode work (records auto-generate these)
-        Cookie duplicate = new Cookie(generatedValue.name(), generatedValue.value(), generatedValue.attributes());
-        assertEquals(generatedValue, duplicate, "Equal cookies should be equal");
-        assertEquals(generatedValue.hashCode(), duplicate.hashCode(), "Equal cookies should have same hash code");
+    private List<String> splitAttributes(String attributes) {
+        if (attributes.isEmpty()) {
+            return List.of();
+        }
+        return Arrays.stream(attributes.split(";")).map(String::trim).toList();
```
[the remark this answers is not in the document]
Comment (OliverWolffGIP) [reply cui-http-core/src/test/java/de/cuioss/http/security/generators/cookie/ValidCookieGeneratorTest.java]: Will be addressed by TASK-10; see follow-up commit on this branch. Fix: isValidAttribute will reject a traversal-marked Path and a negative Max-Age, and splitAttributes will preserve trailing empty fields so a dangling separator fails validation instead of being silently dropped.

File: cui-http-core/src/test/java/de/cuioss/http/security/generators/encoding/PathTraversalGeneratorTest.java [hunk 3]
```diff
@@ -18,25 +18,93 @@
...
+
+        for (int i = 0; i < AGGREGATE_DRAWS; i++) {
+            String value = generator.next();
+            recordIfMatched(attackTypes, "basic", value, BASIC_SIGNATURES);
+            recordIfMatched(attackTypes, "encoded", value, ENCODED_SIGNATURES);
+            recordIfMatched(attackTypes, "double-encoded", value, DOUBLE_ENCODED_SIGNATURES);
+            recordIfMatched(attackTypes, "unicode-literal", value, UNICODE_LITERAL_SIGNATURES);
+            recordIfMatched(attackTypes, "mixed", value, MIXED_SIGNATURES);
+            recordIfMatched(attackTypes, "null-byte", value, NULL_BYTE_SIGNATURES);
+            recordIfMatched(attackTypes, "advanced", value, ADVANCED_SIGNATURES);
```
[the remark this answers is not in the document]
Comment (OliverWolffGIP) [reply cui-http-core/src/test/java/de/cuioss/http/security/generators/encoding/PathTraversalGeneratorTest.java]: Declined - this is a documented, deliberate design decision, not an oversight. The test class javadoc states explicitly that the generator's seven attack types deliberately share encodings (the mixed arm re-emits the basic, encoded and Unicode forms), so the aggregate test asserts reachability of one literal signature per attack type rather than attempting to attribute each value back to the branch that produced it. Making attack-type reachability observable per branch would require exposing new per-branch identifiers on the generator, which is a larger structural change than this plan's scope covers for this generator.

File: cui-http-core/src/test/java/de/cuioss/http/security/generators/GeneratorContractAssertions.java [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+    /**
+     * Truncates a generated value for inclusion in a failure message, so an overlong attack
+     * payload does not flood the test report.
+     *
+     * @param value the value to preview, must not be null
+     * @return {@code value} unchanged when its length is at most 80, otherwise its first 80
+     *         characters followed by {@code "..."}
+     */
+    public static String preview(String value) {
+        return value.length() <= 80 ? value : value.substring(0, 80) + "...";
```
[the remark this answers is not in the document]
Comment (OliverWolffGIP) [reply cui-http-core/src/test/java/de/cuioss/http/security/generators/GeneratorContractAssertions.java]: Will be addressed by TASK-8; see follow-up commit on this branch. Fix: replace the four remaining raw value interpolations in assertContainsAny/assertPipelineAccepts/assertPipelineRejects failure messages with preview(value), matching the truncation already applied elsewhere in this class.

File: cui-http-core/src/test/java/de/cuioss/http/security/generators/injection/URLLengthLimitAttackGeneratorTest.java [hunk 5]
```diff
@@ -15,33 +15,176 @@
...
 
+    /**
+     * Sized so that the rarest branch — the single traversal arm of the algorithmic-complexity
+     * family, reached on one of eight sub-cases of one of thirteen families — is missed with
+     * negligible probability.
+     */
+    private static final int AGGREGATE_DRAWS = 2000;
+
+    /** The number of attack families {@code AttackTypeSelector} cycles through. */
+    private static final int ATTACK_FAMILY_COUNT = 13;
```
[the remark this answers is not in the document]
Comment (OliverWolffGIP) [reply cui-http-core/src/test/java/de/cuioss/http/security/generators/injection/URLLengthLimitAttackGeneratorTest.java]: Will be addressed by TASK-9; see follow-up commit on this branch. Fix: expose public ATTACK_FAMILY_COUNT=13 on URLLengthLimitAttackGenerator and have the test reference it instead of duplicating the literal 13.

File: cui-http-core/src/test/java/de/cuioss/http/security/generators/url/InvalidURLGeneratorTest.java [hunk 6]
```diff
@@ -18,26 +18,123 @@
...
+ *
+ * <p>The defining property of this generator is that every emitted value is either blank or
+ * carries one of the ten documented malformation families. There is deliberately no per-value
+ * pipeline round-trip: the generator emits values such as the well-formed
+ * {@code http://example.com/path/} whose only declared defect is a trailing slash, so a uniform
+ * "the pipeline rejects it" assertion would not hold.</p>
+ *
+ * <p>{@code createPathIssue} case 3 emits a merely trailing-slash URL, which is well-formed
+ * rather than malformed. "Trailing slash" is enumerated below as a declared malformation family
+ * so that this test states honestly what the generator emits today.</p>
```
[the remark this answers is not in the document]
Comment (OliverWolffGIP) [reply cui-http-core/src/test/java/de/cuioss/http/security/generators/url/InvalidURLGeneratorTest.java]: Declined - this is a documented, deliberate design decision, not an oversight. The test class javadoc states explicitly: createPathIssue case 3 emits a merely trailing-slash URL, which is well-formed rather than malformed; 'Trailing slash' is enumerated as a declared malformation family so the test states honestly what the generator emits today. Removing the branch (or the family) would reverse that recorded decision without a corresponding change request in this plan.

File: cui-http-core/src/test/java/de/cuioss/http/security/generators/url/ValidURLPathGeneratorTest.java [hunk 7]
```diff
@@ -15,32 +15,114 @@
...
+    private static final int AGGREGATE_DRAWS = 400;
+
+    private static final Set<String> SYSTEM_PATHS = Set.of("/health", "/metrics", "/status", "/info");
+    private static final String RESOURCES = "(users|orders|products|customers|documents)";
+    private static final Pattern PLAIN_API =
+            Pattern.compile("^/api/" + RESOURCES + "(/\\d+/(search|profile|login|logout))?$");
+    private static final Pattern VERSIONED_API =
+            Pattern.compile("^/api/v[123]/" + RESOURCES + "(/\\d+)?$");
+    private static final Pattern NESTED_RESOURCE =
+            Pattern.compile("^/api/" + RESOURCES + "/\\d+/(items|orders|profile|notifications)$");
```
[the remark this answers is not in the document]
Comment (OliverWolffGIP) [reply cui-http-core/src/test/java/de/cuioss/http/security/generators/url/ValidURLPathGeneratorTest.java]: Will be addressed by TASK-11; see follow-up commit on this branch. Fix: removing the shared profile action from the nested-resource family so a plain-API value can no longer satisfy it by coincidence; NESTED_RESOURCE pattern updated to match. On the ast-grep ReDoS note: false positive - the patterns are compiled once from compile-time-constant string-literal fields (never user input) with no nested/overlapping quantifiers, so none is a catastrophic-backtracking candidate; a comment recording that rationale will be added.

**Comment by OliverWolffGIP** [comment]
## Triage dispositions

### In reply to comment_id: `PRR_kwDOPwSbp88AAAABLHzjoA`

Review summary acknowledged - no independent actionable content beyond the 7 inline comments, all triaged individually above.

File: cui-http-core/src/test/java/de/cuioss/http/security/generators/url/URLLengthLimitAttackGenerator.java [hunk 8]
```diff
@@ -78,6 +90,17 @@
...
+    /**
+     * The number of attack families {@link #next()} cycles through, that is the number of
+     * {@code case} branches its switch declares.
+     *
+     * <p>The former cases 13 and 14 (encoding attacks) were removed because they exercise encoding
+     * validation rather than length validation. This constant is the single source of that count:
+     * it drives the {@link AttackTypeSelector} below, and the contract test references it instead
+     * of duplicating the literal, so adding or removing a family cannot silently desync the two.</p>
+     */
+    public static final int ATTACK_FAMILY_COUNT = 13;
```
[the remark this answers is not in the document]
Comment (OliverWolffGIP) [reply cui-http-core/src/test/java/de/cuioss/http/security/generators/url/URLLengthLimitAttackGenerator.java]: Fixed inline in 0c91b62: ValidURLPathGenerator.PATH_FAMILY_COUNT now bounds pathTypeGen and the contract test asserts the labelled family set size against it. The URLLengthLimitAttackGenerator half was already addressed by ATTACK_FAMILY_COUNT.

**Comment by OliverWolffGIP** [comment]
## Triage dispositions

### In reply to comment_id: `PRR_kwDOPwSbp88AAAABLIgX2g`

Fixed inline in 0c91b62: PURE_PATH_SHAPE now requires isRelative(value), so a hostname-family output can no longer satisfy the pure-path shape.
